libs/constrained-generators/testlib/Test/Min/Model.hs

```diff
@@ -0,0 +1,1306 @@
+
+-- Monitor {} -> pure pred2
+-- Explain es p -> Explain es <$> go fvs sub p
+
+-- ==================================================================================
+
+-- | Lifts 'propagateSpec' to take a Monadic 'Ctx'
+propagateSpecM ::
+  forall v a m.
+  (Monad m, HasSpec v) =>
+  Spec a ->
+  m (Ctx v a) ->
+  m (Spec v)
+propagateSpecM spec ctxM = do ctx <- ctxM; pure $ propagateSpec ctx spec
+
+-- | Precondition: the `Pred` defines the `Var a`
+-- Runs in `GE` in order for us to have detailed context on failure.
+computeSpecSimplified ::
+  forall a. (HasSpec a, HasCallStack) => Var a -> Pred -> GE (Spec a)
+computeSpecSimplified x pred3 = localGESpec $ case simplifyPred pred3 of
+  And ps -> do
+    spec <- fold <$> mapM (computeSpecSimplified x) ps
+    case spec of
+      SuspendedSpec y ps' -> pure $ SuspendedSpec y $ simplifyPred ps'
+      s -> pure s
+  ElemPred True t xs -> propagateSpecM (MemberSpec xs) (toCtx x t)
+  ElemPred False (t :: Term b) xs -> propagateSpecM (TypeSpec @b (emptySpec @b) (NE.toList xs)) (toCtx x t)
+  -- Monitor {} -> pure mempty
+  -- GenHint h t -> propagateSpecM (giveHint h) (toCtx x t)
+  Subst x' t p' -> computeSpec x (substitutePred x' t p') -- NOTE: this is impossible as it should have gone away already
+  TruePred -> pure mempty
+  FalsePred es -> genErrorNE es
+  Let t b -> pure $ SuspendedSpec x (Let t b)
+  Exists k b -> pure $ SuspendedSpec x (Exists k b)
+  Assert (Lit True) -> pure mempty
+  Assert (Lit False) -> genError (show pred3)
+  -- Elem was a pattern
+  -- Assert (Elem _ (Lit [])) -> pure (ErrorSpec (NE.fromList ["Empty list in ElemPat", show pred3]))
+  -- Assert (Elem t (Lit (y : ys))) -> propagateSpecM (MemberSpec (y :| ys)) (toCtx x t)
+  Assert t -> propagateSpecM (equalSpec True) (toCtx x t)
+  ForAll (Lit s) b -> fold <$> mapM (\val -> computeSpec x $ unBind val b) (forAllToList s)
+  ForAll t b -> do
+    bSpec <- computeSpecBinderSimplified b
+    propagateSpecM (fromForAllSpec bSpec) (toCtx x t)
+  Case (Lit val) as bs -> runCaseOn val as bs $ \va vaVal psa -> computeSpec x (substPred (singletonEnv va vaVal) psa)
+  Case t as bs -> do
+    simpAs <- computeSpecBinderSimplified as
+    simpBs <- computeSpecBinderSimplified bs
+    propagateSpecM (typeSpec (SumSpec simpAs simpBs)) (toCtx x t)
+  -- Impossible cases that should be ruled out by the dependency analysis and linearizer
+  DependsOn {} ->
+    fatalErrorNE $
+      NE.fromList
+        [ "The impossible happened in computeSpec: DependsOn"
+        , "  " ++ show x
+        , show $ indent 2 (pretty pred3)
+        ]
+  where
+    -- When (Lit b) tp -> if b then computeSpecSimplified x tp else pure TrueSpec
+    -- This shouldn't happen a lot of the time because when the body is trivial we mostly get rid of the `When` entirely
+    -- When {} -> pure $ SuspendedSpec x pred3
+    {- Reifies (Lit a) (Lit val) f
+      | f val == a -> pure TrueSpec
+      | otherwise ->
+          pure $
```

Comment:
Can we get rid of all the dead code that's been commented out of this module?

Reply:
Yes, it is done